contracts/distribution/subdomains/ZNSSubdomainRegistrar.sol

```diff
@@ -0,0 +1,229 @@
+import { IZNSRegistrar } from "../IZNSRegistrar.sol";
+import { IZNSSubdomainRegistrar } from "./IZNSSubdomainRegistrar.sol";
+import { AAccessControlled } from "../../access/AAccessControlled.sol";
+import { ARegistryWired } from "../../abstractions/ARegistryWired.sol";
+
+
+contract ZNSSubdomainRegistrar is AAccessControlled, ARegistryWired, IZNSSubdomainRegistrar {
+    // TODO sub: change name of Registrar contract
+    IZNSRegistrar public rootRegistrar;
+
+    // TODO sub: make better name AND for the setter function !
+    mapping(bytes32 domainHash => DistributionConfig) public distrConfigs;
+
+    mapping(bytes32 domainHash =>
+        mapping(address registrant => bool allowed)
+    ) public distributionWhitelist;
+
+    modifier onlyOwnerOperatorOrRegistrar(bytes32 domainHash) {
+        require(
+            registry.isOwnerOrOperator(domainHash, msg.sender)
+            || accessController.isRegistrar(msg.sender),
+            "ZNSSubdomainRegistrar: Not authorized"
+        );
+        _;
+    }
+
+    // TODO sub: proxy ??
+    constructor(
+        address _accessController,
+        address _registry,
+        address _rootRegistrar
+    ) {
+        _setAccessController(_accessController);
+        setRegistry(_registry);
+        setRootRegistrar(_rootRegistrar);
+    }
+
+    function registerSubdomain(
+        bytes32 parentHash,
+        string calldata label,
+        address domainAddress,
+        DistributionConfig calldata distrConfig
+    ) external override returns (bytes32) {
+        // TODO sub: make the order of ops better
+        DistributionConfig memory parentConfig = distrConfigs[parentHash];
+        require(
+            parentConfig.accessType != AccessType.LOCKED
+                || registry.isOwnerOrOperator(parentHash, msg.sender),
+            "ZNSSubdomainRegistrar: Parent domain's distribution is locked"
+        );
+
+        if (parentConfig.accessType == AccessType.WHITELIST) {
+            require(
+                distributionWhitelist[parentHash][msg.sender],
+                "ZNSSubdomainRegistrar: Sender is not whitelisted"
+            );
+        }
+
+        bytes32 subdomainHash = hashWithParent(parentHash, label);
+
+        require(
+            !registry.exists(subdomainHash),
+            "ZNSSubdomainRegistrar: Subdomain already exists"
+        );
+
+        uint256 price;
+        uint256 fee;
+        // TODO sub: can we make this abstract switching better ??
+        if (parentConfig.pricingContract.feeEnforced()) {
+            (price, fee) = AZNSPricingWithFee(address(parentConfig.pricingContract)).getPriceAndFee(
+                parentHash,
+                label
+            );
+        } else {
+            price = parentConfig.pricingContract.getPrice(parentHash, label);
+        }
+
+        parentConfig.paymentContract.processPayment(
+            parentHash,
+            subdomainHash,
+            msg.sender,
+            price,
+            fee
+        );
```

Comment:
So if you hit the else on 82, then you've created a local variable for the fee but didn't ever set it and just passed it into processPayment as 0.

The fee logic could be on getPrice(), with the processPayment deciding who to pay with what it's given?

What's the reason they're separate? maybe combine the two contracts?

This is related to the problem below where the feeEnforced() is defined and there's the functions that just return a bool

Reply:
The goal was to make it work with whatever Payment contract. The payment contract then decides what to do with the fee based on it being 0 or not.
If we combine these two contracts, then the modularity and possibility to combine different pricing with different payments goes away. Passing a 0 value in some cases seems like a reasonable sacrifice for the ability to configure.
But I will look into this more, since there might be a way to make it better. Fees are still kind of janky and not finalized. Just wanted to make it work with whatever contract as a start.
Also, `feeEnforced()` call is on the Pricing contract, not payment one. And might be eliminated altogether soon. One of the next PRs after this already remove non-fee pricing contracts.